### PR TITLE
[wip][PoC]Overlay management system

### PR DESCRIPTION
**Pull request by MrsTonedOne**
This is just the basic system, and 1 thing imported in, for public comment.

I have like 20 changed filed in my working directory not commited yet, but I figured I'd put this up for public comment while I work on everything.

The gist is that you can use the new overlay_list as a list, to lower on overhead of adding an overlay, and just apply the list with redraw_overlays().

Every time the overlay list is modified, byond will reconcile the overlay list to its own internal id array and appearance array, as well as start to schedule updates to clients in view. This is a costly operation and by delaying it until all overlays are added/removed we can prevent a lot of that byond overhead in messing with overlays.

The overlay_list has two layers, `OVERLAY_APPEARANCE` and `OVERLAY_PRIORITY`, how ever, (and this is where the fun is) each layer can contain lists and even lists of lists. **Those lists can be keyed**, meaning we can replace all of these per-atom lists of different overlays with the master system.

### CODE REVIEW

File: code/game/over_and_under_lays.dm [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+//	clear_overlays() does not reset priority overlays, the thing that added the overlays is
+//		responsable for removing the individual overlays it added when needed.
+//
+//	clear_overlays() does not call redraw_overlays() or redraw them, you must call redraw_overlays.
+//		(This is so things that do clear->add can delay redraw until after the new overlays are added)
+//
+// Each of the lists can contain their own list of sublists if needed,
+// Associated values are also supported and in those cases, the value is what will be added, not the key
+//		(This means no more storing a seperate list of overlays for the object, you can just key it to a value)
+// 		(You can ofcouse also do a sublists of associated values
```
Comment (RemieRichards) [inline code/game/over_and_under_lays.dm]: ofcouse, I'm unfamiliar with this word and just assume it's a technical term, and not, you know, a typo :P

File: code/game/over_and_under_lays.dm [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+//
+//	(same for underlays of ofcourse)
```
Comment (RemieRichards) [inline code/game/over_and_under_lays.dm]: huh, a variation of the above, I assume (:P)

File: code/game/over_and_under_lays.dm [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+					newoverlays += O
+	overlays = newoverlays
+
+
+//When using this proc, it is still the caller's responsibility to call redraw_underlays().
+//	This keeps byond from sending out two appearance updates to clients needlessly
+//	if the caller then adds underlays.
+/atom/proc/clear_underlays()
+	var/list/newunderlays = OVERLAY_UNDERLAY_LIST
+	newunderlays[UNDERLAY_UNDER_PRIORITY] += underlay_list[UNDERLAY_UNDER_PRIORITY]
```
Comment (RemieRichards) [inline code/game/over_and_under_lays.dm]: Just a question, why do underlays have two priority channels?

**Comment by RemieRichards** [comment]
I like this, and I even like the naming scheme ("channels" has a nice ring to it) which I didn't expect.

File: code/game/over_and_under_lays.dm (hunk 3 above)
Comment (MrStonedOne) [inline code/game/over_and_under_lays.dm]: I changed how this worked mid way thru and forgot to update underlays.

File: code/game/over_and_under_lays.dm (hunk 1 above)
Comment (Shadowlight213) [inline code/game/over_and_under_lays.dm]: It's an American term that the British cannot hope to comprehend, that's why. :^)

Comment (RemieRichards) [inline code/game/over_and_under_lays.dm]: No need to be rude.

**Comment by MrStonedOne** [comment]
its not stale, I've been working on this.

````
C:\Users\kyle\Documents\GitHub\-tg-station [listbasedlays +2 ~41 -0 !]> git diff --stat
 code/__DEFINES/overlay_layers.dm                   |  9 +--
 code/__HELPERS/icon_smoothing.dm                   |  2 +-
 code/__HELPERS/icons.dm                            | 65 +++++++++++++--------
 code/__HELPERS/unsorted.dm                         |  6 +-
 code/_onclick/telekinesis.dm                       |  6 +-
 code/game/atoms.dm                                 |  7 ++-
 code/game/objects/items/crayons.dm                 |  5 +-
 code/game/objects/items/devices/aicard.dm          |  7 ++-
 code/game/objects/items/devices/paicard.dm         | 45 ++++++++------
 code/game/objects/items/devices/taperecorder.dm    |  5 +-
 code/game/objects/items/weapons/defib.dm           | 19 ++++--
 code/game/objects/items/weapons/storage/bags.dm    | 10 +++-
 code/game/objects/items/weapons/storage/belt.dm    |  5 +-
 code/game/objects/items/weapons/storage/fancy.dm   | 36 +++++++-----
 code/game/objects/items/weapons/storage/secure.dm  | 15 +++--
 code/game/over_and_under_lays.dm                   | 46 +++++++--------
 code/game/turfs/open.dm                            |  4 +-
 code/modules/assembly/assembly.dm                  |  5 +-
 .../atmospherics/environmental/LINDA_turf_tile.dm  | 20 +++++--
 code/modules/food_and_drinks/food/customizables.dm | 24 ++++----
 code/modules/food_and_drinks/food/snacks.dm        |  3 +-
 code/modules/holiday/easter.dm                     |  6 +-
 code/modules/mining/mine_turfs.dm                  | 11 ++--
 code/modules/mining/ores_coins.dm                  |  3 +-
 code/modules/mob/dead/observer/observer.dm         | 68 +++++-----------------
 code/modules/mob/living/carbon/human/human.dm      | 11 ++--
 code/modules/mob/living/carbon/update_icons.dm     | 11 ++--
 code/modules/mob/living/living_defense.dm          |  2 +
 code/modules/mob/living/silicon/robot/life.dm      |  7 ++-
 code/modules/mob/living/silicon/robot/robot.dm     | 34 ++++++-----
 .../mob/living/simple_animal/bot/construction.dm   | 20 ++++---
 .../mob/living/simple_animal/friendly/dog.dm       | 23 +++++---
 .../mob/living/simple_animal/hostile/bees.dm       |  9 ++-
 .../mob/living/simple_animal/hostile/mimic.dm      |  7 ++-
 .../living/simple_animal/hostile/mining_mobs.dm    | 11 ++--
 code/modules/paperwork/clipboard.dm                | 15 +++--
 code/modules/projectiles/guns/energy.dm            | 21 +++++--
 code/modules/projectiles/guns/energy/nuclear.dm    |  5 +-
 code/modules/projectiles/guns/energy/special.dm    |  5 +-
 .../projectiles/guns/projectile/automatic.dm       | 29 ++++-----
 tgstation.dme                                      |  1 +
 41 files changed, 361 insertions(+), 282 deletions(-)
C:\Users\kyle\Documents\GitHub\-tg-station [listbasedlays +2 ~41 -0 !]>
````

File: code/__HELPERS/icon_smoothing.dm [hunk 4]
```diff
@@ -226,25 +227,30 @@
...
+	if(!smooth_overlays)
+		smooth_overlays = list()
+		A.overlay_list[A.smooth_overlay_channel]["overlay_smooth"] = smooth_overlays
 
-	if(A.top_right_corner != ne)
-		A.overlays -= A.top_right_corner
-		A.top_right_corner = ne
-		A.add_overlay(ne)
+	var/changed = FALSE
+	if(smooth_overlays["top_left_corner"] != nw)
```
Comment (xxalpha) [inline code/__HELPERS/icon_smoothing.dm]: stop killing icon smoothing

Comment (MrStonedOne) [inline code/__HELPERS/icon_smoothing.dm]: ?

Comment (xxalpha) [inline code/__HELPERS/icon_smoothing.dm]: overhead

Comment (MrStonedOne) [inline code/__HELPERS/icon_smoothing.dm]: Searching for a list key is faster then searching for a datum var in all but certain extreme cases.

Comment (xxalpha) [inline code/__HELPERS/icon_smoothing.dm]: i believe you :white_check_mark:

**Comment by MrStonedOne** [comment]
I'll re-open when i get more time.

**Pull request closed without merging**